src/core/egg-herbie.rkt

```diff
@@ -633,15 +628,28 @@
            (if (hash-has-key? egg->herbie egg-node)
                egg-node             ; variable
                (list egg-node))]    ; constant
-          [(list op child-ids ...)
-           (when (null? child-ids)
-             (vector-set! has-leaf? id #t))
-           (cons op (map resolve-id child-ids))]
+          [(list op child-ids ...) ; application
+           (cond
+             [(null? child-ids) ; application is a constant function
+              (vector-set! has-leaf? id #t)
+              (list op)]
+             [else
+              (define child-ids* (map canon-id child-ids))
+              (for ([child-id (in-list child-ids*)]) ; update parent-child relation
+                (vector-set! parents child-id (cons id (vector-ref parents child-id))))
+              (cons op child-ids*)])]
           [_ (error 'sexpr->regraph "malformed enode: ~a" egg-node)])))
-    (vector-set! eclasses id node))
+    (vector-set! eclasses id nodes))
+
+  ; dedup parent-child relation and convert to vector
+  (for ([id (in-range n)])
+    (vector-set! parents id 
+                 (list->vector
+                   (remove-duplicates
+                     (vector-ref parents id)))))
 
   ; collect with wrapper
-  (regraph eclasses canon has-leaf? constants egg->herbie))
+  (regraph eclasses canon has-leaf? constants parents egg->herbie))
```

Comment:
Does tracking the parent pointers really speed up convergence? Not doubting it per se, but it's a bunch of code.